lib/Dialect/TfheRust/IR/TfheRustDialect.cpp

```diff
@@ -26,6 +29,62 @@ void TfheRustDialect::initialize() {
       >();
 }
 
+// Move an operation as early as possible, so long as it only depends on the
+// server key. This pattern has the downside of adding an annotation to the
+// moved op, but this is necsesary to avoid infinite loops in the pattern
+// matcher. A later pass could remove the attributes, but they are harmless
+// and not emitted in the final codegen.
+template <typename SksOp>
+LogicalResult HoistConstantLikeOps<SksOp>::matchAndRewrite(
+    SksOp op, PatternRewriter &rewriter) const {
+  if (op->hasAttr("hoisted")) {
+    return failure();
+  }
+  DominanceInfo dom(op);
+  Operation *lastOperandDefiner = nullptr;
+  for (Value operand : op->getOperands()) {
```

Comment:
review thoughts: yeah, this seems like the best. i was trying to understand whether you should instead insert at the start of the region required, after considering whether some regions are IsolatedFromAbove and limit how far up you can move the hoistable op.

But if it's an isolated region it's args then the operands would be block args, and you've handled that below.